tools/kdc/kdc-api-server/server/kadmin.go

```diff
@@ -0,0 +1,286 @@
+  }
+
+  // Parse from string
+  err = p.SetFromString(expr)
+  if err != nil {
+    return err
+  }
+
+  return nil
+}
+
+/*
+CreateKAdminClien Create a KAdmin client using the kadmin binary from the given argument
+*/
+func CreateKAdminClient(kadmin string) (*KAdminClient, error) {
+  path, err := exec.LookPath(kadmin)
+  if err != nil {
+    return nil, fmt.Errorf("Unable to lookup kadmin")
+  }
+
+  return &KAdminClient{
+    KAdminPath: path,
+  }, nil
+}
+
+/*
+AddPrincipals Adds one or more principals on KDC
+*/
+func (c *KAdminClient) AddPrincipals(principals []KPrincipal) error {
+  args := []string{
+    "-l",
+    "add",
+    "--use-defaults",
+    "--random-password",
+  }
+  for _, p := range principals {
+    args = append(args, p.Full())
+  }
+
+  // Call-out to KDC
+  _, _, err := c.exec(args...)
+  return err
+}
+
+/*
+AddMissingPrincipals Behaves similarly to AddPrincipals, but it does not throw error if principals already exists
+*/
+func (c *KAdminClient) AddMissingPrincipals(principals []KPrincipal) error {
+  var missingPrincipals []KPrincipal
+
+  // Find missing principals
+  for _, p := range principals {
+    ok, err := c.HasPrincipal(p)
+    if err != nil {
+      return fmt.Errorf("Unable to check if principal '%s' exists: %s", p.Full(), err.Error())
+    }
+
+    // Collect missing
+    if !ok {
+      missingPrincipals = append(missingPrincipals, p)
+    }
+  }
+
+  // Check for empty cases
+  if len(missingPrincipals) == 0 {
+    return nil
+  }
+
+  return c.AddPrincipals(missingPrincipals)
+}
+
+/*
+HasPrincipal Checks if the given principal exists
+*/
+func (c *KAdminClient) HasPrincipal(p KPrincipal) (bool, error) {
+  _, _, err := c.exec("-l", "list", "-l", p.Full())
+  if err != nil {
+    // Check for missing principal
+    if kerr, ok := err.(*KExecError); ok {
+      if strings.Contains(kerr.Stderr, "Principal does not exist") {
+        return false, nil
+      }
+    }
+
+    // Anything else is an error
+    return false, err
+  }
+
+  // If it worked out smoothly, the principal exist
+  return true, nil
+}
+
+/*
+GetKeytab Creates a keytab file for the given principals
+*/
+func (c *KAdminClient) GetKeytabForPrincipals(principals []KPrincipal) ([]byte, error) {
+  // Create a temporary file for the keytab
+  tmpfile, err := ioutil.TempFile("", "keytab")
+  if err != nil {
+    return nil, fmt.Errorf("Unable to create a temporary keytab file")
+  }
+  defer os.Remove(tmpfile.Name()) // clean up
+
+  // Generate keytab
+  args := []string{
+    "-l",
+    "ext",
+    "-k",
+    tmpfile.Name(),
+  }
+  for _, p := range principals {
+    args = append(args, p.Full())
+  }
+
+  // Call-out to KDC
+  _, _, err = c.exec(args...)
+  if err != nil {
+    return nil, err
+  }
+
+  // Collect keytab contents
+  buf := bytes.NewBuffer(nil)
+  _, err = io.Copy(buf, tmpfile)
+  if err != nil {
+    return nil, fmt.Errorf("Unable to read keytab contents: %s", err.Error())
+  }
+
+  // Close the tempfile
```

Comment:
This and many other comments are superfluous, they just rephrase the line of code that follows. Please delete them.